src/librustc_mir/interpret/eval_context.rs

```diff
@@ -1528,6 +1529,36 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
                 Ok(())
             },
             layout::FieldPlacement::Arbitrary { ref offsets, .. } => {
+
+                // check length field and vtable field
+                match layout.ty.builtin_deref(false).map(|tam| &tam.ty.sty) {
+                    | Some(ty::TyStr)
+                    | Some(ty::TySlice(_)) => {
+                        let (len, len_layout) = self.read_field(
+                            Value::ByRef(ptr.into(), ptr_align),
+                            None,
+                            mir::Field::new(1),
+                            layout,
+                        )?;
+                        let len = self.value_to_scalar(ValTy { value: len, ty: len_layout.ty })?;
+                        if len.to_bits(len_layout.size).is_err() {
+                            return validation_failure!("length is not a valid integer", path);
+                        }
+                    },
+                    Some(ty::TyDynamic(..)) => {
+                        let (vtable, vtable_layout) = self.read_field(
+                            Value::ByRef(ptr.into(), ptr_align),
+                            None,
+                            mir::Field::new(1),
+                            layout,
+                        )?;
+                        let vtable = self.value_to_scalar(ValTy { value: vtable, ty: vtable_layout.ty })?;
+                        if vtable.to_ptr().is_err() {
+                            return validation_failure!("vtable address is not a pointer", path);
```

Comment:
In principle, this could even test if it is a vtable, right? (That's what I did for my MIR validation last summer.)

Reply:
Well.. we do go into the vtable because of its layout, but we just do usize tests, which essentially means no unaligned pointers and a big enough allocation.

Do you mean to actually validate the various fields of the vtable?

Reply:
This code just checks if the vtable is a pointer, right? At https://github.com/solson/miri/blob/9d8945e3e8212bf69a54a8f3cb3db3f8dfab1845/src/validation.rs#L746 I have code that makes sure it's a pointer to a vtable (which are special allocations).

Reply:
Here yes, but the layout of the vtable is `[usize; 3+num_methods]`. I could probably do some more sanity testing wrt the values or produce a nicer layout type that does that automatically